core/coreapi/dht.go

```diff
@@ -0,0 +1,241 @@
+
+func (api *DhtAPI) FindPeer(ctx context.Context, p peer.ID) (<-chan ma.Multiaddr, error) {
+	dht, ok := api.node.Routing.(*ipdht.IpfsDHT)
+	if !ok {
+		return nil, ErrNotDHT
+	}
+
+	outChan := make(chan ma.Multiaddr)
+	events := make(chan *notif.QueryEvent)
+	ctx = notif.RegisterForQueryEvents(ctx, events)
+
+	go func() {
+		defer close(outChan)
+
+		sendAddrs := func(responses []*pstore.PeerInfo) error {
+			for _, response := range responses {
+				for _, addr := range response.Addrs {
+					select {
+					case outChan <- addr:
+					case <-ctx.Done():
+						return ctx.Err()
+					}
+				}
+			}
+			return nil
+		}
+
+		for event := range events {
+			if event.Type == notif.FinalPeer {
+				err := sendAddrs(event.Responses)
+				if err != nil {
+					return
+				}
+			}
+		}
+	}()
+
+	go func() {
+		defer close(events)
+		pi, err := dht.FindPeer(ctx, peer.ID(p))
+		if err != nil {
+			notif.PublishQueryEvent(ctx, &notif.QueryEvent{
+				Type:  notif.QueryError,
+				Extra: err.Error(),
+			})
+			return
+		}
+
+		notif.PublishQueryEvent(ctx, &notif.QueryEvent{
+			Type:      notif.FinalPeer,
+			Responses: []*pstore.PeerInfo{&pi},
+		})
+	}()
+
+	return outChan, nil
+}
+
+func (api *DhtAPI) FindProviders(ctx context.Context, p coreiface.Path, opts ...caopts.DhtFindProvidersOption) (<-chan peer.ID, error) {
+	settings, err := caopts.DhtFindProvidersOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	dht, ok := api.node.Routing.(*ipdht.IpfsDHT)
+	if !ok {
+		return nil, ErrNotDHT
+	}
+
+	p, err = api.core().ResolvePath(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	c := p.Cid()
+
+	numProviders := settings.NumProviders
+	if numProviders < 1 {
+		return nil, fmt.Errorf("number of providers must be greater than 0")
+	}
+
+	outChan := make(chan peer.ID)
+	events := make(chan *notif.QueryEvent)
+	ctx = notif.RegisterForQueryEvents(ctx, events)
+
+	pchan := dht.FindProvidersAsync(ctx, c, numProviders)
+	go func() {
+		defer close(outChan)
+
+		sendProviders := func(responses []*pstore.PeerInfo) error {
+			for _, response := range responses {
+				select {
+				case outChan <- response.ID:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+			}
+			return nil
+		}
+
+		for event := range events {
+			if event.Type == notif.Provider {
+				err := sendProviders(event.Responses)
+				if err != nil {
+					return
+				}
+			}
+		}
+	}()
+
+	go func() {
+		defer close(events)
+		for p := range pchan {
+			np := p
+			notif.PublishQueryEvent(ctx, &notif.QueryEvent{
+				Type:      notif.Provider,
+				Responses: []*pstore.PeerInfo{&np},
+			})
+		}
+	}()
+
+	return outChan, nil
+}
+
+func (api *DhtAPI) Provide(ctx context.Context, path coreiface.Path, opts ...caopts.DhtProvideOption) error {
+	settings, err := caopts.DhtProvideOptions(opts...)
+	if err != nil {
+		return err
+	}
+
+	if api.node.Routing == nil {
+		return errors.New("cannot provide in offline mode")
+	}
+
+	if len(api.node.PeerHost.Network().Conns()) == 0 {
```

Comment:
Oof. This is going to hurt. It's fine to do this from the command as we amortize over the request and multiple arguments but listing all the connections just to see if we *have* some connections is going to be very expensive. Ideally, the DHT would detect this and return an error. Does it not do that?

Reply:
This was it the dht command.

Looking at provide code in kad-dht, it invokes `dht.GetClosestPeers`, which I think will error in that case (correct me if I'm wrong) -> https://github.com/libp2p/go-libp2p-kad-dht/blob/2bab49ef85b2d8547e95ca2a699b5be562cc78b6/lookup.go#L60